File: packages/sdk-py/gatewerk/resources/templates.py
"""Templates resource for the Gatewerk Python SDK."""

from __future__ import annotations

from typing import Any, Optional

from .._base import BaseResource
from ..types import Template, TemplateList, TemplateStats


class TemplatesResource(BaseResource):

    def list(self, *, timeout: Optional[float] = None) -> TemplateList:
        data = self._request("GET", "/api/v1/templates", timeout=timeout)
        return TemplateList.model_validate(data)

    def get(self, template_id: str, *, timeout: Optional[float] = None) -> Template:
        data = self._request("GET", f"/api/v1/templates/{template_id}", timeout=timeout)
        return Template.model_validate(data)

    def create(
        self,
        name: str,
        slug: str,
        *,
        description: Optional[str] = None,
        fields: Optional[list[dict[str, Any]]] = None,
        actions: Optional[list[dict[str, Any]]] = None,
        timeout: Optional[float] = None,
    ) -> Template:
        body: dict[str, Any] = {"name": name, "slug": slug}
        if description is not None:
            body["description"] = description
        if fields is not None:
            body["fields"] = fields
        if actions is not None:
            body["actions"] = actions

        data = self._request("POST", "/api/v1/templates", json=body, timeout=timeout)
        return Template.model_validate(data)

    def update(self, template_id: str, *, timeout: Optional[float] = None, **fields: Any) -> Template:
        data = self._request("PUT", f"/api/v1/templates/{template_id}", json=fields, timeout=timeout)
        return Template.model_validate(data)

    def delete(self, template_id: str, *, timeout: Optional[float] = None) -> dict[str, Any]:
        return self._request("DELETE", f"/api/v1/templates/{template_id}", timeout=timeout)

    def stats(self, template_id: str, *, timeout: Optional[float] = None) -> TemplateStats:
        """Get statistics for a specific template."""
        data = self._request("GET", f"/api/v1/templates/{template_id}/stats", timeout=timeout)
        return TemplateStats.model_validate(data)

File: packages/mcp/src/tools/templates.ts
import { z } from "zod";
import type { GatewerkClient } from "gatewerk";
import type { ToolDefinition } from "../types.js";
import { toolSuccess, toolError } from "../types.js";

const fieldSchema = z.object({
  name: z.string(),
  type: z.enum(["text", "markdown", "json", "image", "number", "boolean", "select", "buttons"]),
  label: z.string(),
  readonly: z.boolean().optional(),
  editable: z.boolean().optional(),
  options: z.array(z.string()).optional(),
});

export function templateTools(client: GatewerkClient): ToolDefinition[] {
  return [
    {
      name: "gatewerk_list_templates",
      description: "Discover available review templates. Returns all templates with their fields and actions.",
      scope: "templates:read",
      schema: {},
      handler: async () => {
        const { data, error } = await client.templates.list();
        return error ? toolError(error.message) : toolSuccess(data);
      },
    },
    {
      name: "gatewerk_create_template",
      description: "Define a new review template with form fields and available actions.",
      scope: "templates:write",
      schema: {
        slug: z.string().describe("URL-safe identifier (lowercase, hyphens, e.g., 'invoice-approval')"),
        name: z.string().describe("Human-readable template name"),
        description: z.string().optional().describe("What this template is used for"),
        fields: z.array(fieldSchema).describe("Form fields for the review"),
        actions: z.array(z.string()).describe("Available actions (e.g., ['approve', 'reject', 'edit'])"),
        default_priority: z.enum(["low", "normal", "high", "critical"]).optional(),
      },
      handler: async (params) => {
        const { data, error } = await client.templates.create(params);
        return error ? toolError(error.message) : toolSuccess(data);
      },
    },
    {
      name: "gatewerk_update_template",
      description: "Modify an existing template's name, fields, actions, or priority.",
      scope: "templates:write",
      schema: {
        template_id: z.string().describe("Template ID (gw_tpl_...)"),
        name: z.string().optional(),
        description: z.string().optional(),
        fields: z.array(fieldSchema).optional(),
        actions: z.array(z.string()).optional(),
        default_priority: z.enum(["low", "normal", "high", "critical"]).optional(),
      },
      handler: async (params) => {
        const { template_id, ...updates } = params;
        const { data, error } = await client.templates.update(template_id, updates);
        return error ? toolError(error.message) : toolSuccess(data);
      },
    },
    {
      name: "gatewerk_delete_template",
      description: "Remove a template. Existing reviews using this template are not affected.",
      scope: "templates:write",
      schema: {
        template_id: z.string().describe("Template ID to delete (gw_tpl_...)"),
      },
      handler: async (params) => {
        const { data, error } = await client.templates.delete(params.template_id);
        return error ? toolError(error.message) : toolSuccess(data);
      },
    },
  ];
}
